Support Typespec file for Easm api version 2023-03-01-preview

thang-bit opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

thang-bit committed:
Add typespec file for easm

thang-bit committed:
Change service version doc

thang-bit committed:
Update file structure and add config file

thang-bit committed:
Move examples under a version

thang-bit committed:
Add a openapi.json file

thang-bit committed:
Remove openapi.json file

thang-bit committed:
Change operation names and empty object

thang-bit committed:
Update tspconfig

thang-bit committed:
Change config file to add .net and js emitter

thang-bit committed:
Add service dir to config file

thang-bit committed:
Update csharp config

thang-bit committed:
Add projected name for csharp

thang-bit committed:
Use original op

thang-bit committed:
Update disco group list example

markcowl commented on specification/riskiq/Defender.Easm/main.tsp:
Azure core rules require all of the properties of each model to have documentation, you can do this with `@doc`  decorators, or documentation comments.

In any case, you should try to provide real documentation about what the property contains.  Just adding a comment that gives the name of the property if not sufficient. examples:

```tsp
/** The name of the web component, this name must be unique with the scope fo the application */
name: string
```
or

```tsp
@doc("The name of the web component, this name must be unique with the scope fo the application")
name: string
```

markcowl commented on specification/riskiq/Defender.Easm/main.tsp:
Similarly, both each enum and each enum value must have a documentation comment

thang-bit committed:
Remove observations endpoint

thang-bit committed:
Remove observation models

thang-bit committed:
Add java options

mikeharder commented:
/azp run

thang-bit committed:
Use Azure.Core standard operation and use resource models

thang-bit commented:
@mikekistler, @tg-msft, Can we please have a review at this PR?

catalinaperalta commented on specification/riskiq/Defender.Easm/main.tsp:
```diff
@@ -0,0 +1,1998 @@
+import "@typespec/http";
```
It would be better to separate the routes and models into their own files. So having the routes.tsp and models.tsp to make this code easier to go through :)

catalinaperalta commented:
Also, take a look at the issues reported in the TypeSpec Validation step and address them

thang-bit committed:
Reformat folder structure

thang-bit committed:
Suppress warnings

thang-bit committed:
Change metadata to record<unknown>

shadahmed-microsoft committed:
Add suppression for Record<unknown>

thang-bit committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit committed:
Merge branch 'Azure:main' into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit committed:
Update examples

thang-bit committed:
Merge branch 'riskiq/Micorosft.Easm/2023-03-01-preview' of https://github.com/thang-bit/azure-rest-api-specs into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit committed:
Remove unused examples

thang-bit committed:
Change apiversion to apiVersion

thang-bit committed:
Reformat with prettier

thang-bit committed:
Use foundations for list operation

thang-bit committed:
Update examples and list operations

thang-bit committed:
Remove unused minimum set

thang-bit committed:
Update all uuid

thang-bit committed:
Update Assets example

thang-bit committed:
Run prettier on Assets_List

thang-bit commented:
/azp run

mikeharder commented on specification/riskiq/Easm/tspconfig.yaml:
```diff
@@ -0,0 +1,25 @@
+  "service-dir":
+    default: "sdk/easm"
+emit:
+  - "@azure-tools/typespec-apiview"
```
```suggestion
```

Remove this emitter

mikeharder replied:
@thang-bit: This is probably caused by mismatched versions of TypeSpec.  Follow the steps here to ensure you are using the latest version:

https://github.com/Azure/azure-rest-api-specs/pull/24937#pullrequestreview-1566171892

thang-bit replied:
I'm getting this error when running the final step `npx tsv path/to/my/spec`:
`SyntaxError: The requested module 'node:util' does not provide an export named 'parseArgs'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:128:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:194:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:385:24)
    at async loadESM (node:internal/process/esm_loader:88:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12)`
I'm using node version 16.15.1

mikeharder replied:
I believe you need to upgrade to `16.17.0` or higher:

https://nodejs.org/en/blog/release/v16.17.0#notable-changes

Thank you for reporting this, we will update the minimum required node versions in the `tsv` tool.

thang-bit replied:
The typespec validation test passed. Thank you, Mike. Can you share with me the tsv tool documentation?

mikeharder replied:
The tool does not have documentation yet, but you can view the sources here: https://github.com/Azure/azure-rest-api-specs/tree/main/eng/tools/TypeSpecValidation

mikeharder reviewed (commented):
To fix check `TypeSpec Validation`, after making my suggested change to your `tspconfig.yaml`, , please run the following commands, review the changes, and push the updated files to your PR.

```
$ cd azure-rest-api-specs
$ git checkout your-branch
$ git merge main
$ npm ci
$ npx tsv path/to/your/spec
```

catalinaperalta commented on specification/riskiq/Easm/main.tsp:
```diff
@@ -0,0 +1,42 @@
+using Azure.Core;
+using TypeSpec.Versioning;
+
+@useAuth(OAuth2Auth<[ImplicitFlow]>)
```
I think you need to define this like in the example here: https://github.com/Azure/typespec-azure/blob/907bd205e66943dc3d5e89947d8de9d826f0dbf4/packages/samples/specs/data-plane/languageqna/Interface/main.tsp#L7-L13

@markcowl is the definition I linked to in the sample the correct way to define an implicit auth flow?

shadahmed-microsoft committed:
update oauth2 definition

shadahmed-microsoft committed:
remove implicitflow model

catalinaperalta commented on specification/riskiq/Easm/main.tsp:
```diff
@@ -0,0 +1,51 @@
+    title: "Defender EASM",
+})
+@server(
+    "{endpoint}/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/workspaces/{workspaceName}",
```
Why is the server url defined like this? You most likely just want `{endpoint}` like it was defined in the original swagger.

catalinaperalta replied:
Nevermind, I just noticed what the routes look like, it seems better to specify these parameters once at the server level instead of on each operation

catalinaperalta commented on specification/riskiq/Easm/models.tsp:
```diff
@@ -0,0 +1,1635 @@
+    ...SkipQueryParameter;
+    ...MaxPageSizeQueryParameter;
+}>;
+alias SavedFiltersListTrait = QueryParametersTrait<{
```
This alias seems to be have the same parameters as the one on line 26 and 37, I'd suggest just defining one with a generic name and reusing it on the corresponding routes. Same feedback for the others.

shadahmed-microsoft committed:
Update specification/riskiq/Easm/tspconfig.yaml

Co-authored-by: Mike Harder <mharder@microsoft.com>

shadahmed-microsoft committed:
Update models.tsp

shadahmed-microsoft committed:
Update routes.tsp

thang-bit committed:
Remove typespec-apiview emitter

thang-bit committed:
Merge branch 'riskiq/Micorosft.Easm/2023-03-01-preview' of https://github.com/thang-bit/azure-rest-api-specs into riskiq/Micorosft.Easm/2023-03-01-preview

shadahmed-microsoft committed:
update examples, add client.tsp

thang-bit committed:
run prettier

thang-bit committed:
Run prettier on Easm examples

thang-bit commented:
/azp run

thang-bit committed:
Update examples, main.tsp. Remove client.tsp

thang-bit committed:
Add operation id to examples under typespec project

thang-bit committed:
Add titles to examples

thang-bit committed:
Reformat typespec files

thang-bit committed:
Update DataConnection validate example

thang-bit committed:
Update data connection validation example

thang-bit committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit committed:
Update the easm.json file based on tsv tool

thang-bit commented:
/azp run

mikekistler commented on specification/riskiq/Easm/models.tsp:
I thought we had decided to use "Discovery" rather than "Disco".

thang-bit replied:
We have been using this naming pattern for the paths, and this version has been approved. I would keep this model naming pattern to keep it consistent with the path

mikekistler commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,166 @@
+  @doc("Retrieve a data connection with a given dataConnectionName.")
+  getDataConnection is Operations.ResourceRead<DataConnection>;
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or update a data connection with a given dataConnectionName.")
```
HTTP "put" method is "create or replace".
```suggestion
  @doc("Create or replace a data connection with a given dataConnectionName.")
```

mikekistler commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,166 @@
+  @doc("Retrieve a saved filter by filterName.")
+  getSavedFilter is Operations.ResourceRead<SavedFilter>;
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or update a saved filter with a given filterName.")
```
HTTP put is create or replace
```suggestion
  @doc("Create or replace a saved filter with a given filterName.")
```

mikekistler commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,166 @@
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or update a saved filter with a given filterName.")
+  @put
+  putSavedFilter is Foundations.ResourceOperation<
```
Don't use the http method in the operation name -- describe what the operation does (just like in the description).
```suggestion
  createOrReplaceSavedFilter is Foundations.ResourceOperation<
```

mikekistler commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,166 @@
+  getDataConnection is Operations.ResourceRead<DataConnection>;
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or update a data connection with a given dataConnectionName.")
+  @put
+  putDataConnection is Foundations.ResourceOperation<
+    DataConnection,
+    DataConnectionData,
+    DataConnection
+  >;
```
This operation does not follow REST conventions. The request body should a "representation of the resource" and the response should be an "equivalent representation of the resource".  But:
- "name" is passed in the body and as a path parameter, so what value is set for "name" in the response?
- There is no "id" in the request body but it appears in the response body.
- There is no "displayName" in the request body but it appears in the response body.
- The response contains several "metadata" properties that should be marked "readOnly": "createdDate", "updatedDate", "userUpdatedAt", "active", "inactiveMessage"

mikekistler commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,166 @@
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create a discovery group with a given groupName.")
+  @put
+  putDiscoGroup is Foundations.ResourceOperation<
```
```suggestion
  createOrReplaceDiscoGroup is Foundations.ResourceOperation<
```

mikekistler replied:
A createOrReplace operation should return 200 and 201.

mikekistler reviewed (changes requested):
Please address these comments, not only where I have specifically commented but generally in the API, and then tag me to re-review.

thang-bit committed:
Update swagger file and its examples

thang-bit committed:
Update Easm typespec examples

thang-bit committed:
Update uris in examples

thang-bit committed:
Update cvss3 name

thang-bit committed:
Update cvss3 model with tsp

mikekistler reviewed (approved):
This is acceptable for an initial preview but needs some important changes / improvements that should be done in a new preview.

thang-bit committed:
Add totalElements to DiscoGroups, DiscoTemplates, and SavedFilters List traits

thang-bit committed:
Run validation for the previous commit

thang-bit commented:
/azp run

mikekistler commented:
The Swagger Breaking Changes do not affect the actual REST API -- only how it is being described. There a no generated SDKs yet so these changes will not cause customer disruption.

thang-bit committed:
Suppress linting rule of typespec generated swagger

thang-bit commented:
Hi @markcowl, it seems that the PR merging is blocked because of existing requested changes. We aren't able to put comments on all the properties in this version, but we put comments to address them in the next release

weidongxu-microsoft commented on specification/riskiq/Easm/tspconfig.yaml:
```diff
@@ -0,0 +1,24 @@
+      name: "@azure-rest/defender-easm-rest"
+  "@azure-tools/typespec-java":
+    package-dir: "azure-defender-easm"
+    namespace: com.azure.defender.easm
```
namespace should be what you have in arch review https://github.com/Azure/azure-sdk/issues/5062 `com.azure.analytics.defender.easm`
https://github.com/Azure/azure-sdk-for-java/pull/36282/files#r1300780400

thang-bit replied:
@weidongxu-microsoft I have updated the namspaces using the approved names

thang-bit committed:
Update namespace for dotnet and javascript to the approved namespace

thang-bit committed:
Update package directory for java sdk from azure-defender-easm to azure-analytics-defender-easm

mikekistler commented:
@markcowl Could you please re-review and approve or indicate what changes are still needed?

thang-bit committed:
use AlexaDetails instead of AlexaInfo in java sdk

thang-bit commented:
/azp run

thang-bit committed:
Merge branch 'Azure:main' into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit commented:
Hi @mikeharder, I'm getting TypeSpec Validation after pushing a change to the PR. Then, I tried to fix by running these commands

```
cd azure-rest-api-specs
git merge main
npm ci 
npx tsv specification/riskiq/Easm
```
I'm getting this error when running the last command: 
```
Executing rule: FolderStructure
Verify tspconfig file extension
Rule FolderStructure failed
Please limit TypeSpec folder depth to 3 levels or lessInvalid folder structure. Package must contain main.tsp or client.tsp, tspconfig.yaml, and examples folder if there's main.tsp.

```
According to this error, should I change the current folder structure?

Thank you

mikeharder commented:
> Hi @mikeharder, I'm getting TypeSpec Validation after pushing a change to the PR. Then, I tried to fix by running these commands
> 
> ```
> cd azure-rest-api-specs
> git merge main
> npm ci 
> npx tsv specification/riskiq/Easm
> ```
> 
> I'm getting this error when running the last command:
> 
> ```
> Executing rule: FolderStructure
> Verify tspconfig file extension
> Rule FolderStructure failed
> Please limit TypeSpec folder depth to 3 levels or lessInvalid folder structure. Package must contain main.tsp or client.tsp, tspconfig.yaml, and examples folder if there's main.tsp.
> ```
> 
> According to this error, should I change the current folder structure?
> 
> Thank you

Are you seeing this error locally?  If so, can you please share more logs showing how I can reprodfuce the error locally?  I don't see this error in the "TypeSpec Validation" check in your PR.

thang-bit commented:
This is the error logs when I run `npx tsv`
<img width="1293" alt="image" src="https://github.com/Azure/azure-rest-api-specs/assets/127871101/eba1f7af-8923-4b0c-9877-a2f6d048451b">

Also, is this the correct way to solve the error that I'm getting?

mikeharder commented:
> This is the error logs when I run `npx tsv` <img alt="image" width="1293" src="https://user-images.githubusercontent.com/127871101/272401098-eba1f7af-8923-4b0c-9877-a2f6d048451b.png">
> 
> Also, is this the correct way to solve the error that I'm getting from "Typespec Validations"?

I think you want to run `npx tsv specification/riskiq/Easm` (without the leading slash).  Separately I think we can improve this error message in TSV, it's misleading because the folder doesn't exist at all on your machine.

thang-bit committed:
Fix git diff issue in Typespec Validation by running npx tsp

thang-bit commented:
Ah I see, removing the backlash works. Thank you

markcowl commented on specification/riskiq/Easm/tspconfig.yaml:
```diff
@@ -0,0 +1,24 @@
+    azure-resource-provider-folder: "data-plane"
+    emitter-output-dir: "{project-root}/.."
+    examples-directory: "examples"
+    output-file: "{azure-resource-provider-folder}/Microsoft.Easm/preview/2023-03-01-preview/easm.json"
```
You should use the variables for version and version type, like this:

```yaml
output-file: "{azure-resource-provider-folder}/Microsoft.Easm/{version-status}/{version}/easm.json"
```

as shown here: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml#L18

markcowl commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,172 @@
+  @doc("Get asset summary details for the summary request.")
+  @route("/reports/assets:getSummary")
+  @post
+  getSummary is Foundations.Operation<
```
Why can't you use the resource action templates here?

thang-bit replied:
When I use `Operations.ResourceCollectionAction`, it would result in `/assets:getSummary` instead of `/reports/assets:getSummary`

thang-bit commented on specification/riskiq/Easm/models.tsp:
```diff
@@ -0,0 +1,1678 @@
+alias AssetsListTrait = {
+  ...PageResultBodyFields<{
+    @doc("The total number of items available in the full result set.")
+    totalElements?: int64;
```
So, some of the endpoints don't use all of these query parameters, so I think it'd make sense if each list request have different its own list trait

thang-bit commented on specification/riskiq/Easm/models.tsp:
```diff
@@ -0,0 +1,1678 @@
+}
+
+@doc("The page of assets that match the provided metric.")
+model AssetPageResult {
```
This `AssetPageResult` model is only used as `assets` property of `ReportAssetSnapshotResult` model, can we use trait as a child model of other models?

thang-bit commented on specification/riskiq/Easm/models.tsp:
```diff
@@ -0,0 +1,1678 @@
+  recent?: boolean;
+}
+
+#suppress "@azure-tools/typespec-azure-core/documentation-required" "Documentation will be added next preview version"
```
Given the sheer number of properties in this specification file, we are working on process of documenting these properties. For the current specification version, it was recommended for us by the review board that we should suppress the warning and add comment to make sure that we will document all of these properties in the next version.

thang-bit commented on specification/riskiq/Easm/models.tsp:
Remove this `DiscoGroupPageResult` since it's not used in any operation

thang-bit commented on specification/riskiq/Easm/models.tsp:
Remove this `DiscoTemplatePageResult` since it's not used in any operation

thang-bit commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,172 @@
+  @doc("Retrieve a list of data connections.")
+  listDataConnection is Operations.ResourceList<
+    DataConnection,
+    DataConnectionsListTrait
```
Thanks for pointing this out. I don't really need all of those custom paged results. So, I remove them and use the list trait in the operation instead

thang-bit commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,172 @@
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Update labels on assets matching the provided filter.")
+  @route("/assets")
+  @post
```
This is a POST action to update labels on asset resources. I'm not sure why this route doesn't have the `:action` moniker but this is the pattern that have been approved by reviewers from multiple reviews. Therefore, I decided to use Foundation Operation here

thang-bit commented on specification/riskiq/Easm/models.tsp:
This would make the models.tsp file shorter, but I feel like having both of them is more intuitive when looking at the models

thang-bit committed:
Merge branch 'main' of https://github.com/thang-bit/azure-rest-api-specs into riskiq/Micorosft.Easm/2023-03-01-preview

thang-bit committed:
Remove unused paged result models, add @pagedResult, @nextLink, and @items decorators to paged result models and their properties.
Remove custom NoContentResult model and use Typespec.Http.NoContentResponse model instead

markcowl commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,169 @@
+  getDataConnection is Operations.ResourceRead<DataConnection>;
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or replace a data connection with a given dataConnectionName.")
+  @put
```
This should add a decorator to indicate that is an update or createOrUpdate  or createOrReplace Operation for DataConnection,

likely `@TypeSpec.Rest.createsOrReplacesResource(DataConnection)`

markcowl commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,169 @@
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create a discovery group with a given groupName.")
+  @put
+  createOrReplaceDiscoGroup is Foundations.ResourceOperation<
```
This should add a decorator to indicate that is an update or createOrUpdate  or createOrReplace Operation for DiscoGroup,

likely `@TypeSpec.Rest.createsOrReplacesResource(DiscoGroup)`

markcowl commented on specification/riskiq/Easm/routes.tsp:
```diff
@@ -0,0 +1,169 @@
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @doc("Create or replace a saved filter with a given filterName.")
+  @put
+  createOrReplaceSavedFilter is Foundations.ResourceOperation<
```
This should add a decorator to indicate that is an update or createOrUpdate  or createOrReplace Operation for SavedFilter,

likely `@TypeSpec.Rest.createsOrReplacesResource(SavedFilter)`

markcowl reviewed (changes requested):
SHould add rest operation decorators for resource operations that did not fit into the templates.  This preserves the relationship of this operation to a lifecycle operation for the resource.

I also don't understand the documentation suppressions - if you are checking this spec in, I assume this is for generating clients, and the documentation you put on models will be transferred to documentation in the generated clients

thang-bit committed:
Add decorators to indicate the resource is createOrUpdate or just update
